Sources/SourceKitLSP/Swift/Refactoring.swift

```diff
@@ -0,0 +1,129 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+
+protocol RefactoringResponse {
+  init(title: String, uri: DocumentURI, refactoringEdits: [RefactoringEdit])
+}
+
+extension RefactoringResponse {
+  /// Create an instance of `RefactoringResponse` from a sourcekitd semantic
+  /// refactoring response dictionary, if possible.
+  ///
+  /// - Parameters:
+  ///   - title: The title of the refactoring action.
+  ///   - dict: Response dictionary to extract information from.
+  ///   - snapshot: The snapshot that triggered the `semantic_refactoring` request.
+  ///   - keys: The sourcekitd key set to use for looking up into `dict`.
+  init?(_ title: String, _ dict: SKDResponseDictionary, _ snapshot: DocumentSnapshot, _ keys: sourcekitd_api_keys) {
+    guard let categorizedEdits: SKDResponseArray = dict[keys.categorizedEdits] else {
+      logger.fault("categorizedEdits doesn't exist in response dictionary")
+      return nil
+    }
+
+    var refactoringEdits = [RefactoringEdit]()
+
+    categorizedEdits.forEach { _, categorizedEdit in
+      guard let edits: SKDResponseArray = categorizedEdit[keys.edits] else {
+        logger.fault("edits doesn't exist in categorizedEdit dictionary")
+        return true
+      }
+      edits.forEach { _, edit in
+        guard let startLine: Int = edit[keys.line],
+          let startColumn: Int = edit[keys.column],
+          let endLine: Int = edit[keys.endLine],
+          let endColumn: Int = edit[keys.endColumn],
+          let text: String = edit[keys.text]
+        else {
+          logger.fault("Failed to deserialise edit dictionary containing values: \(edit)")
+          return true  // continue
+        }
+
+        // The LSP is zero based, but semantic_refactoring is one based.
+        let startPosition = snapshot.positionOf(
+          zeroBasedLine: startLine - 1,
+          utf8Column: startColumn - 1
+        )
+        let endPosition = snapshot.positionOf(
+          zeroBasedLine: endLine - 1,
+          utf8Column: endColumn - 1
+        )
+        // Snippets are only supported in code completion.
+        // Remove SourceKit placeholders in refactoring actions because they
+        // can't be represented in the editor properly.
+        let textWithSnippets = rewriteSourceKitPlaceholders(in: text, clientSupportsSnippets: false)
+        refactoringEdits.append(
+          RefactoringEdit(
+            range: startPosition..<endPosition,
+            newText: textWithSnippets,
+            bufferName: edit[keys.bufferName] ?? ""
```

Comment:
Do you think a cleaner design would be to make `RefactoringEdit.bufferName` an optional? Then we wouldn’t need to nil-coalesce here and I think `nil` represents the absence of a buffer name better than an empty string.

Reply:
Sure, and I would also highlight the fact that it is ok for bufferName to be nil (i.e. not passed as a response from sourcekitd) if the refactoring edit is meant to represent the removal of the macro annotation from the original source file (here, newText will be empty). Its only a fault of sourcekitd if the refactoring edit's buffer name is nil when the refactoring edit represents an expansion or replacement and has an associated newText value which won't be empty.